README
Gnuk - An Implementation of USB Cryptographic Token for GnuPG

							    Version 2.1
							     2022-0?-??
							   Niibe Yutaka
				      Free Software Initiative of Japan

Release Notes
=============

This is the release of Gnuk, version 2.1, which has major incompatible
changes to Gnuk 1.

Please update your documentation for Gnuk Token, so that the
instruction of importing keys won't cause any confusion.

It has supports of Ed25519 and X25519 (ECDH on Curve25519).  It also
has experimental support of ECDSA on secp256k1 and ECDH on secp256k1.

It supports new KDF-DO feature.  Please note that this is
experimental.  To use the feature, you need to use newer GnuPG (2.2.6
or later).  You need to prepare the KDF-DO on your token by the
card-edit/kdf-setup command.

With FST-01SZ and GNU/Linux emulation, experimental ack button support
is available for test.


What's Gnuk?
============

Gnuk is an implementation of USB cryptographic token for GNU Privacy
Guard.  Gnuk supports OpenPGP card protocol version 3, and it runs on
STM32F103 processor (and its compatible).

I wish that Gnuk will be a developer's soother who uses GnuPG.  I have
been nervous of storing secret key(s) on usual secondary storage.
There is a solution with OpenPGP card, but it is not the choice for
me, as card reader is not common device.  With Gnuk, this issue will
be solved by a USB token.

Please look at the graphics of "gnuk.svg" for the software name.  My
son used to be with his NUK(R), always, everywhere.  Now, I am with a
USB Cryptographic Token by "Gnuk", always, everywhere.


FAQ
===

Q0: How Gnuk USB Token is superior than other solutions (OpenPGP
    card 2.0/3.3/3.4, YubiKey, etc.) ?
    https://www.g10code.de/p-card.html
    https://www.yubico.com/
A0: Good points of Gnuk are:
    * If you have skill of electronics and like DIY, you can build
      Gnuk Token cheaper (see Q8-A8).
    * You can study Gnuk to modify and to enhance.  For example, you
      can implement your own authentication method with some sensor
      such as an acceleration sensor.
    * It is "of Free Software"; Gnuk is distributed under GPLv3+,
	    "by Free Software"; Gnuk development requires only Free Software
				(GNU Toolchain, Python, etc.), 
	    "for Free Software"; Gnuk supports GnuPG.

Q1: What kind of key algorithm is supported?
A1: Gnuk version 1.0 only supports RSA-2048.
    Gnuk version 1.2.x supports 255-bit EdDSA, as well as RSA-4096.
    (Note that it takes long time to sign with RSA-4096.)

Q2: How long does it take for digital signing?
A2: It takes a second and a half or so for RSA-2048. 
    It takes more than 8 seconds for RSA-4096.

Q3: What's your recommendation for target board?
A3: Orthodox choice is Olimex STM32-H103.
    FST-01SZ (Flying Stone Tiny 01 SZ) is available for sale, and it
    is a kind of the best choice, hopefully.  If you have a skill of
    electronics, STM32 Nucleo F103 is the best choice for experiment.

Q4: What's version of GnuPG are you using?
A4: In Debian GNU/Linux system, I use GnuPG modern 2.2.23.

Q5: What's version of pcscd and libccid are you using?
A5: I don't use them, pcscd and libccid are optional, you can use Gnuk
    Token without them.
    I tested pcscd 1.5.5-4 and libccid 1.3.11-2 which were in Debian
    squeeze.

Q6: What kinds of hardware is required for development?
A6: You need a target board plus a JTAG/SWD debugger.  If you just
    want to test Gnuk for target boards with DfuSe, JTAG debugger is
    not the requirement.  Note that for real use, you need JTAG/SWD
    debugger to enable flash ROM protection.

Q7: How much does it cost?
A7: Olimex STM32-H103 plus ARM-USB-TINY-H cost 70 Euro or so.

Q8: How much does it cost for DIY version?
A8: STM32 Nucleo F103 costs about $10 USD.

Q9: I got an error like "gpg: selecting openpgp failed: ec=6.108", what's up?
A9: Older GnuPG's SCDaemon has problems for handling insertion/removal of
    card/reader.  When your newly inserted token is not found by
    GnuPG, try killing scdaemon and let it to be invoked again.  I do:

	 $ gpg-connect-agent "SCD KILLSCD" "SCD BYE" /bye

    and confirm scdaemon doesn't exist, then,

	 $ gpg-connect-agent learn /bye

Qa: With GNOME 2, I can't use Gnuk Token for SSH.  How can we use it for SSH?
Aa: You need to deactivate seahorse-agent and gnome-keyring, but use
    gpg-agant for the role of ssh-agent.  For gnome-keyring please do:

      $ gconftool-2 --type bool --set /apps/gnome-keyring/daemon-components/ssh false

Qb: With GNOME 3.0, I can't use Gnuk Token at all.  Why?
Ab: That's because gnome-keyring-daemon interferes GnuPG.  Type:

      $ gnome-session-properties

    and at the tab of "Startup Programs", disable check buttons for
    "GPG Password Agent" and "SSH Key Agent".

Qc: With GNOME 3.x (x >= 8?), I can't use Gnuk Token at all.  Why?
Ac: That's because gnome-keyring-daemon interferes GnuPG.  Please
    disable the invocation of gnome-keyring-daemon.  In Debian
    wheezy, it's in the files /etc/xdg/autostart/gnome-keyring-ssh.desktop
    and /etc/xdg/autostart/gnome-keyring-gpg.desktop.
    We have a line something like:

        OnlyShowIn=GNOME;Unity;MATE;

    Please edit this line to:

        OnlyShowIn=

Qd: Do you know a good SWD debugger to connect FST-01 or something?
Ad: ST-Link/V2 is cheap one.  We have a tool/stlinkv2.py as flash ROM
    writer program.  STM32 Nucleo F103 comes with the valiant of
    ST-Link/V2.  Note that the firmware of ST-Link/V2 is proprietary.
    So, in case of transparency matters, ST-Link/V2 would not be your
    choice.
    I care transparency for our process of manufacturing FST-01SZ (and
    better control by Free Software, in general), thus, I develop
    BBG-SWD, SWD debugger by BeagleBone Green.
    I use ST-Link/V2 for daily development.  For serious task like
    flashing product, I use BBG-SWD.


Tested features
===============

Gnuk is tested by test suite.  Please see the "tests" directory.

	* Personalization of the card
	  * Changing Login name, URL, Name, Sex, Language, etc.
	* Password handling (PW1, RC, PW3)
	* Key import for three types:
	  * key for digital signing
	  * key for decryption
	  * key for authentication
	* PSO: Digital Signature
	* PSO: Decipher
	* INTERNAL AUTHENTICATE
	* Changing value of password status bytes (0x00C4): forcesig
	* Verify with pin pad
	* Modify with pin pad
	* Card holder certificate (read)
	* Removal of keys
	* Key generation on device side for RSA-2048
	* Overriding key import

Original features of Gnuk, tested manually lightly:

	* OpenPGP card serial number setup
	* Card holder certificate (write by UPDATE BINARY)
	* Upgrading with "EXTERNAL AUTHENTICATE" by reGNUal


Targets
=======

We use Olimex STM32-H103 board and Flying Stone Tiny 01 (FST-01).

With DfuSe support, STBee is also our targets.  But this target with
DfuSe is for experiment only, because it is impossible for DfuSe to
disable read from flash.  For real use, please consider killing DfuSe
and enabling read protection using JTAG debugger.


Build system and Host system
============================

Makefile is written for GNU make.  You need Bash 4.x for configure.

If your bash is not installed as /bin/bash, you need to run configure
script prepending 'bash' before './configure'.

Some tools are written in Python.  If your Python is not installed as
/usr/bin/python, please prepend 'python' or 'python3' for your command
invocation.  I use Python 3.8 and PyUSB 1.0.2.


Source code
===========

Gnuk source code is under src/ directory.

Note that SHA-2 hash function implementation, src/sha256.c, is based
on the original implementation by Dr. Brian Gladman.  See:

  http://brg.a2hosted.com//oldsite/cryptography_technology/sha/index.php
(was at:
 http://gladman.plushost.co.uk/oldsite/cryptography_technology/sha/index.php)


License
=======

It is distributed under GNU General Public Licence version 3 or later
(GPLv3+).  Please see src/COPYING.

Please note that it is distributed with external source code too.
Please read relevant licenses for external source code as well.

The author(s) of Gnuk expect users of Gnuk will be able to access the
source code of Gnuk, so that users can study the code and can modify
if needed.  This doesn't mean person who has a Gnuk Token should be
able to access everything on the Token, regardless of its protections.
Private keys, and other information should be protected properly.


External source code
====================

Gnuk is distributed with external source code.

* chopstx/  -- Chopstx 2.4

  We use Chopstx as the kernel for Gnuk.

  Chopstx is distributed under GPLv3+ (with a special exception).


USB vendor ID and product ID (USB device ID)
============================================

When you have a vendor ID and assign a product ID for Gnuk, edit the
file GNUK_USB_DEVICE_ID and add an entry for yours.  In this case,
please contact Niibe, so that it is listed to the file in the official
release of the source code.

When you are modifing Gnuk and installing the binary to device, you
should replace the vendor string and serial number to yours (in the
file GNUK_USB_DEVICE_ID and SERIALNO of the script of src/configure),
so that users can see it's not by original vendor, and it is modified
version.

FSIJ allows you to use USB device ID of FSIJ (234b:0000) for devices
with Gnuk under one of following conditions:

  * For everyone for experimental purpose:

    - You must not distribute a binary with FSIJ's USB device ID, but
      must use the binary by yourself only for your experiment.  Note
      that "Distributing binary" includes distributing a device which
      holds the binary.

  * For general individuals:

    - You must use your Gnuk device with a card serial number which is
      *not* by FSIJ.  Easy one would be a card serial number generated
      by chip unique ID.

  * For individuals with explicit permission from FSIJ.

    - You should have an assigned card serial number by FSIJ,
      please use that number for your device.
      (There a file 'GNUK_SERIAL_NUMBER' in the official release.)

FSIJ could give companies or business entities "second source
manufacturer" license to use USB device ID of FSIJ for devices with
unmodified version of Gnuk, provided they support Free Software and
respect users' freedom for computing.  Please ask FSIJ for the
license.

Otherwise, companies which want to distribute Gnuk devices, please use
your own USB vendor ID and product ID.  Please replace vendor string
and possibly product string to yours, when you modify Gnuk.


Host Requirements
=================

For GNU/Linux, PC/SC service is an option, you can use GnuPG's
internal CCID driver instead.  If you chose using PC/SC service,
libccid version >= 1.3.11 is recommended for GNU/Linux.


How to compile
==============

You need GNU toolchain and newlib for 'arm-none-eabi' target.

On Debian we can install the packages of gcc-arm-none-eabi
and its friends.  I'm using:

	binutils-arm-none-eabi	2.37-7+15
	gcc-arm-none-eabi 	15:10.3-2021.07-4
	libnewlib-arm-none-eabi	3.3.0-1
	gdb-multiarch 		10.1-2

Or else, see https://launchpad.net/gcc-arm-embedded for preparation of
GNU Toolchain for 'arm-none-eabi' target.

Change directory to `src':

  $ cd gnuk-VERSION/src

Then, run `configure':

  $ ./configure --vidpid=<VID:PID>

Here, you need to specify USB vendor ID and product ID.  For FSIJ's,
it's: --vidpid=234b:0000 .  Please read section 'USB vendor ID and
product ID' above.


Then, type:

  $ make

Then, we will have "gnuk.elf" under src/build directory.

If you are not the authorized vendor, please never distribute this
file of "gnuk.elf", which includes VID:PID in the image.  If you would
like to distribute the image (for example, to check if it's
reproducible or not), the file "gnuk-no-vidpid.elf" is the one with no
VID:PID.


How to install
==============

Olimex STM32-H103 board
-----------------------

If you are using Olimex JTAG-Tiny, type following to invoke OpenOCD
and write "gnuk.elf" to Flash ROM:

  $ openocd -f interface/ftdi/olimex-jtag-tiny.cfg \
            -f board/olimex_stm32_h103.cfg \
            -c "program build/gnuk.elf verify reset exit"

Command invocation is assumed in src/ directory.


Flying Stone Tiny 01
--------------------

If you are using Flying Stone Tiny 01, you need a SWD writer.

OpenOCD 0.9.0 now supports ST-Link/V2.  We can use it like:

  $ openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg \
            -c "program build/gnuk.elf verify reset exit"


STBee
-----

Note that this is only for your experiment; Your private key materials
on the board can be accessed by DfuSe.

Reset the board with "USER" switch pushed.  Type following to write
to flash:

  # cd ../tool
  # ./dfuse.py ../src/build/gnuk.hex

Then, reset the board.


How to protect flash ROM
========================

To protect, invoke OpenOCD like (for FST-01):

  $ openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg \
            -c init -c "reset halt" -c "stm32f1x lock 0" -c reset -c exit

After power-off / power-on sequence, the contents of flash ROM cannot
be accessible from JTAG debugger.

Unprotecting is:

  $ openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg \
            -c init -c "reset halt" -c "stm32f1x unlock 0" -c reset -c exit

Upon unprotection, flash is erased.

Note that it would be still possible for some implementation of DfuSe
to access the contents, even if it's protected.  If you really want to
protect, killing DfuSe and accessing by JTAG debugger is recommended.


(Optional) Configure serial number and X.509 certificate
========================================================

This is completely optional.


NEWS
Gnuk NEWS - User visible changes


* Major changes in Gnuk 2.1

  Released 2022-XX-XX, by NIIBE Yutaka

** Removal of RSA support

** Replace AES implementation

** Change of authentication for firmware upgrade


* Major changes in Gnuk 1.2.19

  Released 2021-10-12, by NIIBE Yutaka

** KDF Data Object configuration
KDF Data Object should be highly recommended for all configurations.
Nevertheless, for backward compatibillity, in Gnuk 1.2, it is optional
by default; It is up to user to configure KDF Data Object before
importing private keys.  In this situation, it is not good to
introduce new build-time option like --enable-always-require-kdf-do,
because it might wrongly encourage use of Gnuk with no KDF Data Object
setting, by confusion.  If needed, please run configure:

  kdf_do=required ./configure --enable-factory-reset --target...

or

  kdf_do=optional ./configure --enable-factory-reset --target...

Please note that such a use of variable by shell command line is not
well supported by the configure script (for other variables), but
override of kdf_do is needed in some situations.

** Upgrade of Chopstx
We use Chopstx 1.20.  This enables use with PC/SC for GNU/Linux
emulation.


* Major changes in Gnuk 1.2.18

  Released 2021-04-02, by NIIBE Yutaka

** GNU/Linux emulation bug fix
This time, for GNU/Linux emulation, KDF Data Object is required before
keygen and key import, really.

** tool/upgrade_by_passwd.py
Now, it checks if the target of binary being written and the
configured target of running device are same.  This check can be
skipped by -s option.  Please note that FST-01 and FST-01SZ are
different target, as it's MHz is different.


* Major changes in Gnuk 1.2.17

  Released 2021-02-25, by NIIBE Yutaka

** GNU/Linux: KDF Data Object is required before keygen and key import
For GNU/Linux emulation, KDF Data Object is required before keygen and
key import.

** GNU/Linux emulation
Since 1.2.10, it was not build-able because of MHZ definition.  It is
build-able again, and its USB product string is now "Gnuk Token".  It
has ACK button support using terminal.  It now has start-up message,
and its driver show useful information to setup USBIP.  When no file
for the .gnuk-flash-image file, it is automatically created.

** Removal of tool/gnuk-emulation-setup
Because it is automatically created, the tool is not needed any more.

** Upgrade of Chopstx
We use Chopstx 1.19.


* Major changes in Gnuk 1.2.16

  Released 2020-09-10, by NIIBE Yutaka

** New Data Object (Algorithm Information) of OpenPGP card v3.4
The tag is 0x00FA.  This is useful for user interaction to show which
algorithms are supported by the device.

** Ed25519 signing allowing longer message
For OpenPGP, it does hashing on host side before requesting signing to
the device.  Thus, the length of message to be signed is limited and
determined by the hash algorithm.  That's good feature of OpenPGP.  On
the other hand, there is a use case, like OpenSSH certificate signing,
where the length of message is a kind of arbitrary.  Even though Gnuk
(or OpenPGP card protocol itself) has limitation, we removed the
length check against EDDSA_HASH_LEN_MAX at cmd_pso.


* Major changes in Gnuk 1.2.15

  Released 2020-01-24, by NIIBE Yutaka

** Switch to Python3
Scripts under tool/ are switched to Python3.
Thanks to Bertrand Jacquin.

** Upgrade of Chopstx
We use Chopstx 1.18.

** Tests also support OpenPGPcard
Now, a test suite under "tests" may be used to OpenPGPcard.


* Major changes in Gnuk 1.2.14

  Released 2019-03-05, by NIIBE Yutaka

** Timeout for ACK button support
When a user doesn't acknowledge (> 15 seconds), the operation
timeouts, and authentication state is cleared.

** Upgrade of Chopstx
We use Chopstx 1.14.


* Major changes in Gnuk 1.2.13

  Released 2018-12-26, by NIIBE Yutaka

** DFU support and its firmware upgrade fix
DFU support was not well maintained, and firmware upgrade was not
possible for boards with DFU.  Now, at least for Maple Mini, it is
tested.  Note that using Gnuk with DFU on a board is only for an
experiment, because DFU can access the content of flash ROM.  DFU
should be killed by upgrading to normal Gnuk, so that you can have
your private keys.

** Fix for UIF Data Object
When flash ROM is full and coping to new page, UIF DO was not properly
copied.  This bug resulted losing the flag for user interaction.  Now,
it's properly copied, keeping the setting of the feature.

** Upgrade of Chopstx
We use Chopstx 1.13.


* Major changes in Gnuk 1.2.12

  Released 2018-11-25, by NIIBE Yutaka

** FST-01SZ fixes
Fixes for Ack button support and serial number.


* Major changes in Gnuk 1.2.11

  Released 2018-11-12, by NIIBE Yutaka

** Experimental ACK button support with FST-01SZ
While OpenPGP card specification verison 3 has description for the
"user interaction" button and data objects, there were no
implementation.  To evaluate the feature, experimental support is
added.

** Upgrade of Chopstx
We use Chopstx 1.12, which comes with ACK button driver and supports
FST-01SZ.


* Major changes in Gnuk 1.2.10

  Released 2018-05-10, by NIIBE Yutaka

** No inclusion of VID:PID in gnuk-no-vidpid.elf
Now, we have new file named gnuk-no-vidpid.elf with no VID:PID.  The
file gnuk.elf has the VID:PID, like version 1.2.7 or older.

** Upgrade of Chopstx
We use Chopstx 1.9, which supports GD32F103.


* Major changes in Gnuk 1.2.9

  Released 2018-04-05, by NIIBE Yutaka

** A test suite fix: Clear PW3
Until 1.2.8, after running the test suite under "tests", PW3 keystring
remained, which affects use of admin-less mode.  New test case is
added to clear PW3.

** tool/upgrade_by_passwd.py supports KDF-DO auth
With KDF-DO, firmware upgrade didn't work.  Now, it's supported.

** Add "single-salt" support for KDF-DO
With KDF-DO, "admin-less" mode didn't work well.  With new feature of
"single-salt" support, we can use "admin-less" mode with KDF-DO.

** factory-reset deletes all upgrade public keys

src/main.c
/*
 * main.c - main routine of Gnuk
 *
 * Copyright (C) 2010, 2011, 2012, 2013, 2015, 2016, 2017, 2018, 2021,
 *               2022
 *               Free Software Initiative of Japan
 * Author: NIIBE Yutaka <gniibe@fsij.org>
 *
 * This file is a part of Gnuk, a GnuPG USB Token implementation.
 *
 * Gnuk is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Gnuk is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#include <stdint.h>
#include <string.h>
#include <chopstx.h>
#include <eventflag.h>

#include "config.h"

#include "sys.h"
#include "adc.h"
#include "gnuk.h"
#include "usb_lld.h"
#include "usb-cdc.h"
#include "random.h"
#ifdef GNU_LINUX_EMULATION
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include <stdio.h>
#include <stdlib.h>
#define main emulated_main
#else
#include "mcu/stm32f103.h"
#endif

/*
 * main thread does 1-bit LED display output
 */
#define LED_TIMEOUT_INTERVAL	(75*1000)
#define LED_TIMEOUT_ZERO	(25*1000)
#define LED_TIMEOUT_ONE		(100*1000)
#define LED_TIMEOUT_STOP	(200*1000)

#ifdef DFU_SUPPORT
static int
flash_write_any (uintptr_t dst_addr, const uint8_t *src, size_t len)
{
  int status;

  while (len)
    {
      uint16_t hw = *src++;

      hw |= (*src++ << 8);
      status = flash_program_halfword (dst_addr, hw);
      if (status != 0)
        return 0;              /* error return */

      dst_addr += 2;
      len -= 2;
    }

  return 1;
}
#endif

#ifdef GNU_LINUX_EMULATION
uint8_t *flash_addr_key_storage_start;
uint8_t *flash_addr_data_storage_start;
#else
#define ID_OFFSET (2+SERIALNO_STR_LEN*2)
static void
device_initialize_once (void)
{
  const uint8_t *p = &gnuk_string_serial[ID_OFFSET];

  if (p[0] == 0xff && p[1] == 0xff && p[2] == 0xff && p[3] == 0xff)
    {
      /*
       * This is the first time invocation.
       * Setup serial number by unique device ID.
       */
      const uint8_t *u = unique_device_id () + (MHZ < 96 ? 8: 0);
      int i;

      for (i = 0; i < 4; i++)
	{
	  uint8_t b = u[3-i];
	  uint8_t nibble;

	  nibble = (b >> 4);
	  nibble += (nibble >= 10 ? ('A' - 10) : '0');
	  flash_put_data_internal (&p[i*4], nibble);
	  nibble = (b & 0x0f);
	  nibble += (nibble >= 10 ? ('A' - 10) : '0');
	  flash_put_data_internal (&p[i*4+2], nibble);
	}

#ifdef DFU_SUPPORT
#define CHIP_ID_REG ((uint32_t *)0xE0042000)
      /*
       * Overwrite DFU bootloader with a copy of SYS linked to ORIGIN_REAL.
       * Then protect flash from readout.
       */
      {
        extern uint8_t _binary_build_stdaln_sys_bin_start;
        extern uint8_t _binary_build_stdaln_sys_bin_size;
        size_t stdaln_sys_size = (size_t) &_binary_build_stdaln_sys_bin_size;
        extern const uint32_t FT0[256], FT1[256], FT2[256];
        extern handler vector_table[];
        uintptr_t addr;
        uint32_t flash_page_size = 1024; /* 1KiB default */

        if (((*CHIP_ID_REG)&0x07) == 0x04) /* High density device.  */
          flash_page_size = 2048; /* It's 2KiB. */

        /* Kill DFU */
        for (addr = ORIGIN_REAL; addr < ORIGIN;
             addr += flash_page_size)
          flash_erase_page (addr);

        /* Copy SYS */
        addr = ORIGIN_REAL;
        flash_write_any(addr, &_binary_build_stdaln_sys_bin_start,
                        stdaln_sys_size);
        addr += stdaln_sys_size;
        flash_write_any(addr, (const uint8_t *) &FT0, sizeof(FT0));
        addr += sizeof(FT0);
        flash_write_any(addr, (const uint8_t *) &FT1, sizeof(FT1));
        addr += sizeof(FT1);
        flash_write_any(addr, (const uint8_t *) &FT2, sizeof(FT2));

        addr = ORIGIN_REAL + 0x1000;
        if (addr < ORIGIN) {
          /* Need to patch top of stack and reset vector there */
          handler *new_vector = (handler *) addr;
          flash_write((uintptr_t) &new_vector[0], (const uint8_t *)
                      &vector_table[0], sizeof(handler));
          flash_write((uintptr_t) &new_vector[1], (const uint8_t *)
                      &vector[1], sizeof(handler));
        }

        flash_protect();
        nvic_system_reset();
      }
#endif
    }
}
#endif


static volatile uint8_t fatal_code;
static struct eventflag led_event;
static chopstx_poll_cond_t led_event_poll_desc;
static struct chx_poll_head *const led_event_poll[] = {
  (struct chx_poll_head *)&led_event_poll_desc
};

static void display_fatal_code (void)
{
  while (1)
    {
      set_led (1);
      chopstx_usec_wait (LED_TIMEOUT_ZERO);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_INTERVAL);
      set_led (1);
      chopstx_usec_wait (LED_TIMEOUT_ZERO);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_INTERVAL);
      set_led (1);
      chopstx_usec_wait (LED_TIMEOUT_ZERO);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_STOP);
      set_led (1);
      if (fatal_code & 1)
	chopstx_usec_wait (LED_TIMEOUT_ONE);
      else
	chopstx_usec_wait (LED_TIMEOUT_ZERO);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_INTERVAL);
      set_led (1);
      if (fatal_code & 2)
	chopstx_usec_wait (LED_TIMEOUT_ONE);
      else
	chopstx_usec_wait (LED_TIMEOUT_ZERO);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_INTERVAL);
      set_led (1);
      chopstx_usec_wait (LED_TIMEOUT_STOP);
      set_led (0);
      chopstx_usec_wait (LED_TIMEOUT_INTERVAL*10);
    }
}

static uint8_t led_inverted;

static void
emit_led (uint32_t on_time, uint32_t off_time)
{
  set_led (!led_inverted);
  chopstx_poll (&on_time, 1, led_event_poll);
  set_led (led_inverted);
  chopstx_poll (&off_time, 1, led_event_poll);
}

static void
display_status_code (void)
{
  enum ccid_state ccid_state = ccid_get_ccid_state ();
  uint32_t usec;

  if (ccid_state == CCID_STATE_START)
    emit_led (LED_TIMEOUT_ONE, LED_TIMEOUT_STOP);
  else
    /* OpenPGP card thread is running */
    {
      emit_led ((auth_status & AC_ADMIN_AUTHORIZED)?
		LED_TIMEOUT_ONE : LED_TIMEOUT_ZERO, LED_TIMEOUT_INTERVAL);
      emit_led ((auth_status & AC_OTHER_AUTHORIZED)?
		LED_TIMEOUT_ONE : LED_TIMEOUT_ZERO, LED_TIMEOUT_INTERVAL);
      emit_led ((auth_status & AC_PSO_CDS_AUTHORIZED)?
		LED_TIMEOUT_ONE : LED_TIMEOUT_ZERO, LED_TIMEOUT_INTERVAL);

      if (ccid_state == CCID_STATE_WAIT)
	{
	  usec = LED_TIMEOUT_STOP * 2;
	  chopstx_poll (&usec, 1, led_event_poll);
	}
      else
	{
	  usec = LED_TIMEOUT_INTERVAL;
	  chopstx_poll (&usec, 1, led_event_poll);
	  emit_led (LED_TIMEOUT_ZERO, LED_TIMEOUT_STOP);
	}
    }
}

void
led_blink (int spec)
{
  if (spec == LED_START_COMMAND || spec == LED_FINISH_COMMAND)
    {
      led_inverted = (spec == LED_START_COMMAND);
      spec = LED_SYNC;
    }

  eventflag_signal (&led_event, spec);
}

#ifdef FLASH_UPGRADE_SUPPORT
/*
 * In Gnuk 1.0.[12], reGNUal was not relocatable.
 * Now, it's relocatable, but we need to calculate its entry address
 * based on it's pre-defined address.
 */
#define REGNUAL_START_ADDRESS_COMPATIBLE 0x20001400
static uintptr_t
calculate_regnual_entry_address (const uint8_t *addr)
{
  const uint8_t *p = addr + 4;
  uintptr_t v = p[0] + (p[1] << 8) + (p[2] << 16) + (p[3] << 24);

  v -= REGNUAL_START_ADDRESS_COMPATIBLE;
  v += (uintptr_t)addr;
  return v;
}
#endif

#define STACK_MAIN
#define STACK_PROCESS_1
#include "stack-def.h"
#define STACK_ADDR_CCID ((uintptr_t)process1_base)
#define STACK_SIZE_CCID (sizeof process1_base)

#define PRIO_CCID 3
#define PRIO_MAIN 5

extern void *ccid_thread (void *arg);

extern uint32_t bDeviceState;

/*
 * Entry point.
 */
int
main (int argc, const char *argv[])
{
#ifdef GNU_LINUX_EMULATION
  uintptr_t flash_addr;
  const char *flash_image_path;
  char *path_string = NULL;
#endif
#ifdef FLASH_UPGRADE_SUPPORT
  uintptr_t entry;
#endif
  chopstx_t ccid_thd;
  int wait_for_ack = 0;

  chopstx_conf_idle (1);

#ifdef GNU_LINUX_EMULATION
#define FLASH_IMAGE_NAME ".gnuk-flash-image"

  if (argc >= 4 || (argc == 2 && !strcmp (argv[1], "--help")))
    {
      fprintf (stdout, "Usage: %s [--vidpid=Vxxx:Pxxx] [flash-image-file]",
	       argv[0]);
      exit (0);
    }

  if (argc >= 2 && !strncmp (argv[1], "--debug=", 8))
    {
      debug = strtol (&argv[1][8], NULL, 10);
      argc--;
      argv++;
    }

  if (argc >= 2 && !strncmp (argv[1], "--vidpid=", 9))
    {
      extern uint8_t device_desc[];
      uint32_t id;
      char *p;

      id = (uint32_t)strtol (&argv[1][9], &p, 16);
      device_desc[8] = (id & 0xff);
      device_desc[9] = (id >> 8);

      if (p && p[0] == ':')
	{
	  id = (uint32_t)strtol (&p[1], NULL, 16);
	  device_desc[10] = (id & 0xff);
	  device_desc[11] = (id >> 8);
	}

      argc--;
      argv++;
    }

  if (argc == 1)
    {
      char *p = getenv ("HOME");

      if (p == NULL)
	{
	  fprintf (stderr, "Can't find $HOME\n");
	  exit (1);
	}

      path_string = malloc (strlen (p) + strlen (FLASH_IMAGE_NAME) + 2);

      p = stpcpy (path_string, p);
      *p++ = '/';
      strcpy (p, FLASH_IMAGE_NAME);
      flash_image_path = path_string;
    }
  else
    flash_image_path = argv[1];

  if (access (flash_image_path, F_OK) < 0)
    {
      int fd;
      char buf[8192];

      memset (buf, 0xff, sizeof buf);
      memset (buf+4*1024, 0, 2);
      fd = open (flash_image_path, O_CREAT|O_WRONLY, S_IRUSR|S_IWUSR);
      if (fd < 0)
	{
	  perror ("creating flash file");
	  exit (1);
	}

      if (write (fd, buf, sizeof buf) != sizeof buf)
	{
	  perror ("initializing flash file");
	  close (fd);
	  exit (1);
	}

      close (fd);
    }

  puts ("Gnuk (emulation with USBIP), a GnuPG USB Token implementation");
  puts ("Copyright (C) 2021 Free Software Initiative of Japan");
  puts ("This is free software under GPLv3+.");

  flash_addr = flash_init (flash_image_path);
  flash_addr_key_storage_start = (uint8_t *)flash_addr;
  flash_addr_data_storage_start = (uint8_t *)flash_addr + 4096;
#else
  (void)argc;
  (void)argv;
#endif

  flash_unlock ();

#ifdef GNU_LINUX_EMULATION
    if (path_string)
      free (path_string);
#else
  device_initialize_once ();
#endif

  adc_init ();

  eventflag_init (&led_event);

  random_init ();

#ifdef DEBUG
  stdout_init ();
#endif

  ccid_thd = chopstx_create (PRIO_CCID, STACK_ADDR_CCID, STACK_SIZE_CCID,
			     ccid_thread, NULL);

  chopstx_setpriority (PRIO_MAIN);

  while (1)
    {
      if (bDeviceState != USB_DEVICE_STATE_UNCONNECTED)
	break;

      chopstx_usec_wait (250*1000);
    }

  eventflag_prepare_poll (&led_event, &led_event_poll_desc);

  while (1)
    {
      eventmask_t m;

      if (wait_for_ack)
	m = eventflag_wait_timeout (&led_event, LED_TIMEOUT_INTERVAL);
      else
	m = eventflag_wait (&led_event);

      switch (m)
	{
	case LED_ONESHOT:
	  emit_led (100*1000, LED_TIMEOUT_STOP);
	  break;
	case LED_TWOSHOTS:
	  emit_led (50*1000, 50*1000);
	  emit_led (50*1000, LED_TIMEOUT_STOP);
	  break;
	case LED_SHOW_STATUS:
	  display_status_code ();
	  break;
	case LED_FATAL:
	  display_fatal_code ();
	  break;
	case LED_SYNC:
	  set_led (led_inverted);
	  break;
	case LED_GNUK_EXEC:
	  goto exec;
	case LED_WAIT_FOR_BUTTON:
	  wait_for_ack ^= 1;
	  /* fall through */
	default:
	  emit_led (LED_TIMEOUT_ZERO, LED_TIMEOUT_ZERO);
	  break;
	}
    }

 exec:
  random_fini ();

  set_led (1);
  usb_lld_shutdown ();

  /* Finish application.  */
  chopstx_join (ccid_thd, NULL);

#ifdef FLASH_UPGRADE_SUPPORT
  /* Set vector */
  SCB->VTOR = (uintptr_t)_regnual_start;
  entry = calculate_regnual_entry_address (_regnual_start);
#ifdef DFU_SUPPORT
  {
    /* Use SYS at ORIGIN_REAL instead of the one at ORIGIN */
    handler *new_vector = (handler *)ORIGIN_REAL;
    void (*func) (void (*)(void)) = (void (*)(void (*)(void))) new_vector[9];

    /* Leave Gnuk to exec reGNUal */
    (*func) ((void (*)(void))entry);
    for (;;);
  }
#else
  /* Leave Gnuk to exec reGNUal */
  flash_erase_all_and_exec ((void (*)(void))entry);
#endif
#else
  exit (0);
#endif

  /* Never reached */
  return 0;
}

void
fatal (uint8_t code)
{
  extern void _write (const char *s, int len);

  fatal_code = code;
  eventflag_signal (&led_event, LED_FATAL);
  _write ("fatal\r\n", 7);
  for (;;);
}
